Istio native sidecar blog (#4334)
Azure/AKS: blog/_posts/2024-06-06-istio-native-sidecar.md

date: 2024-06-06
author: Fuyuan Bie
categories:
  - networking
  - addons
---

We're excited to announce the public preview of native sidecar support for Istio add-on for Azure Kubernetes Service (AKS). This blog post will guide you through the fundamentals of Kubernetes native sidecars, the integration of native sidecar support in Istio add-on, the benefits of this feature, how to enable it, and a demo showcasing its advantages.

## Understanding Kubernetes Native Sidecars

In Kubernetes, the sidecar pattern is used to extend the functionality of applications running in pods. Sidecar containers are responsible for auxiliary tasks such as logging, proxying, and monitoring. Typically, managing sidecar lifecycle required special configurations, which could be complex and error-prone. Developers must understand the lifecycle interactions of containers within a Kubernetes pod to avoid unexpected behaviors. They cannot safely assume that sidecars will be ready before their application or expect a specific shutdown order. Additionally, for jobs, the primary job container must terminate the sidecars.

> Starting from Kubernetes v1.28, native sidecar support becomes a builtin feature. On AKS, starting from v1.29, this feature is turned on.

Reviewer: let's remove the k8s 1.28 part, it was in alpha in 1.28. for AKS we start consuming features as it gets to beta. we can instead mention that native sidecars reached beta in k8s upstream 1.29 and that AKS as a result has it too starting 1.29 now

Read more about this feature from [Kubernetes blog](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/).

## Native Sidecar Support in Istio 1.20+

With the release of Istio 1.20, native sidecar support has been introduced, offering a streamlined approach to managing sidecar proxies. This integration reduces the overhead and complexity associated with sidecar injection, providing efficient management of sidecar life cycles.

Read more about Istio native sidecar support from [Istio blog](https://istio.io/latest/blog/2023/native-sidecars/).

## Where Native Sidecar Does not Help

> While native sidecar support addresses many challenges, it does not assist with egress traffic as much as with ingress traffic. This is because within the lifecycle of a pod, application containers are terminated after sidecar containers, leaving no effective way to notify the application container that the pod is being terminated before the native sidecar is stopped. Consequently, in-flight egress traffic from the application container may experience disruptions during pod restarts.

Reviewer: We are presenting the gap on where it doesn't help. do we have any recommendation to customer on how to handle it? or can we add a note on how this compares with non-native sidecar model?

Author: Explicitly called out the service developers still have to expect disruptions and handle them properly.

## Try it out!

### Enabling Istio Native Sidecar Support in AKS

> Since this is currently a preview feature, we need to follow [these instructions](https://learn.microsoft.com/en-us/azure/aks/istio-native-sidecar#before-you-begin) to register `IstioNativeSidecarModePreview` preview feature.

Reviewer 2: Removing locale from MS learn URL.

```suggestion
Since this is currently a preview feature, we need to follow [these instructions](https://learn.microsoft.com/azure/aks/istio-native-sidecar#before-you-begin) to register `IstioNativeSidecarModePreview` preview feature.
```

   Check sidecar logs.

   ```bash
   POD_ID=$(kubectl get pod -l app=httpbin -o jsonpath='{.items[0].metadata.name}')
   kubectl logs $POD_ID -c istio-proxy
   ```

   Sidecar container istio-proxy used 9 seconds to get ready.

   ```text
   2024-06-04T05:07:02.138613Z    info    cache    returned workload trust anchor from cache    ttl=23h59m59.861388298s
   2024-06-04T05:07:10.904779Z    info    Readiness succeeded in 9.015247842s
   2024-06-04T05:07:10.905752Z    info    Envoy proxy is ready
   ```

   Events show application container was created 10 seconds after istio-proxy container was created. That is 1 second after the istio-proxy sidecar became ready.

   | LAST SEEN | TYPE | REASON | OBJECT | MESSAGE |
   |-----------|------|--------|--------|---------|
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-proxy |
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-init |
   | 2m28s | Normal | Created | pod/httpbin-9c5fdf746-hzmsf | Created container httpbin |
   | 2m28s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container httpbin |

>    At the termination time, sidecar get killed before the application container.  This gives sidecar time to drain inflight traffic before terminating the application container.

Reviewer: ```suggestion
   At the termination time, sidecar gets killed before the application container.  This gives sidecar time to drain inflight traffic before terminating the application container.
```

   2024-06-04T05:36:59.579521Z    info    Agent already drained, exiting immediately
   2024-06-04T05:36:59.579525Z    info    Agent has successfully terminated
   ```
   
   From events, we can see application container is killed after istio-proxy sidecar.

   | LAST SEEN | TYPE | REASON | OBJECT | MESSAGE |
   |-----------|------|--------|--------|---------|
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container istio-proxy |
   | 2m27s | Normal | SuccessfulDelete | replicaset/httpbin-598f48c6d | Deleted pod: httpbin-598f48c6d-nwx67 |
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container httpbin |

## Conclusion

The integration of native sidecar support in Istio 1.20 and its inclusion in the AKS Istio addon marks a significant advancement in simplifying and enhancing service mesh deployment and management. This feature provides substantial benefits, including improved performance, simplified configuration, and operational efficiency. We are committed to providing state-of-the-art features that empower our users to build resilient and scalable applications on AKS.

## Credits

> This feature was made possible through the collaborative efforts of the AKS traffic team, the AKS Add-on team, and the Istio community. Special thanks to Niranjan Shankar, Shashank Barsin and all other AKS traffic team members for the collaboration; Robbie Zhang for the help given from AKS CCP and Add-on team; and Brian Redmond for his guidance on this blog and last but not least Paul Yu for the review.

Reviewer: these are internal (addon team, traffic) team and will not be relevant for users. we can send an internal mail thanking these folks. suggest taking out this section to align with format of other posts on the blog

Author: Not quite sure.  At a common practice, most tech blogs from other companies carries acknowledgements section for other teams or team members who helped. That is a great way to express gratitude.

---
title: "Introducing Native Sidecar Support for Istio addon on AKS"

> description: "This blog talks about native sidecar support for Istio addon on AKS."

Reviewer 2: Improving the meta description to make it more SEO friendly.

```suggestion
description: "Explore native sidecar support for the Istio add-on in Azure Kubernetes Service (AKS). Learn its benefits, how to enable it, and see a practical demo."
```

date: 2024-06-06
author: Fuyuan Bie

> categories: mesh

Reviewer 2: ```suggestion
categories:
  - networking
  - addons
```

---

> ## Overview

Reviewer 2: This heading is probably not necessary so removing this sub-heading to reduce spacing between page header and post content.

```suggestion
## Overview
```

```bash
az feature register --namespace Microsoft.ContainerService --name IstioNativeSidecarModePreview
```

It will take a few minutes to finish registration. Run the following command to check registration status:

```bash
az feature show --namespace Microsoft.ContainerService --name IstioNativeSidecarModePreview
```

Once the registration status shows `Registered`, refresh Azure feature registration for AKS.

```bash
az provider register --namespace Microsoft.ContainerService
```

### Create an AKS cluster with Istio add-on

Create a new AKS cluster and connect to it. Kubernetes 1.29+ is required in order to enable the native sidecar feature.

```bash
az group create \
  --name native_sidecar_test_rg \
  --location eastus

az aks create \
  --resource-group native_sidecar_test_rg \
  --name native_sidecar_test \
  --kubernetes-version 1.29 \
  --enable-asm \
  --revision asm-1-21 \
  --location eastus

az aks get-credentials \
  --resource-group native_sidecar_test_rg \
  --name native_sidecar_test \

>   --overwrite-existing

Author: Hmm. How could it work when I tested.

>   --overwrite-existing

Reviewer 2: Not sure. I'm using Azure CLI version 2.61 with aks-preview extension version 3.0.0b13 and got this error message when testing:

```
--name cannot exceed 63 characters and can only contain letters, numbers, or dashes (-).
```

```

### Deploy a sample application

Let's deploy a sample application to test Istio native sidecar.
```bash
kubectl label namespace default istio.io/rev=asm-1-21 --overwrite
kubectl apply -f https://raw.githubusercontent.com/istio/istio/master/samples/httpbin/httpbin.yaml
```

Check application pod:

```bash
kubectl get pods -o "custom-columns=NAME:.metadata.name,INIT:.spec.initContainers[*].name,CONTAINERS:.spec.containers[*].name"
```

We can see `istio-proxy` now becomes an init container. A native sidecar is an init container.

> ```log

Reviewer 2: `text` renders better than `log`

```suggestion
```text
```

   | 2m39s | Normal | ScalingReplicaSet | deployment/httpbin | Scaled up replica set httpbin-9c5fdf746 to 1 |
   | 2m38s | Normal | Created | pod/httpbin-9c5fdf746-hzmsf | Created container istio-init |
   | 2m38s | Normal | Created | pod/httpbin-9c5fdf746-hzmsf | Created container istio-proxy |
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-proxy |
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-init |
   | 2m28s | Normal | Created | pod/httpbin-9c5fdf746-hzmsf | Created container httpbin | 
   | 2m28s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container httpbin |
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container istio-proxy |
   | 2m27s | Normal | SuccessfulDelete | replicaset/httpbin-598f48c6d | Deleted pod: httpbin-598f48c6d-nwx67 |
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container httpbin |

   We can see, when the pod starts, istio-proxy as an init container starts before the application container. It gives enough time for istio-proxy to warm up before serving traffic.

>    istio-proxy used 9 seconds to get ready.

Reviewer 2: Seems that the following kubectl command is missing here to see the logs below.

```bash
POD_ID=$(kubectl get pod -l app=httpbin -o jsonpath='{.items[0].metadata.name}')
kubectl logs $POD_ID -c istio-proxy
```

>    ```log

Reviewer 2: `text` renders better than `log`

```suggestion
   ```text
```

   2024-06-04T05:07:02.138613Z    info    cache    returned workload trust anchor from cache    ttl=23h59m59.861388298s
   2024-06-04T05:07:10.904779Z    info    Readiness succeeded in 9.015247842s
   2024-06-04T05:07:10.905752Z    info    Envoy proxy is ready
   ```

   Events show application container was created 10 seconds after istio-proxy container was created. That is 1 second after the istio-proxy sidecar became ready.

   | LAST SEEN | TYPE | REASON | OBJECT | MESSAGE |
   |-----------|------|--------|--------|---------|
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-proxy |
   | 2m38s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container istio-init |
   | 2m28s | Normal | Created | pod/httpbin-9c5fdf746-hzmsf | Created container httpbin |
   | 2m28s | Normal | Started | pod/httpbin-9c5fdf746-hzmsf | Started container httpbin |

   At the termination time, sidecar get killed before the application container.  This gives sidecar time to drain inflight traffic before terminating the application container.

>    ```log

Reviewer 2: `text` renders better than `log`. Also, is there a command the reader should run to see these logs?

```suggestion
   ```text
```

   |-----------|------|--------|--------|---------|
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container istio-proxy |
   | 2m27s | Normal | SuccessfulDelete | replicaset/httpbin-598f48c6d | Deleted pod: httpbin-598f48c6d-nwx67 |
   | 2m27s | Normal | Killing | pod/httpbin-598f48c6d-nwx67 | Stopping container httpbin |

## Conclusion

The integration of native sidecar support in Istio 1.20 and its inclusion in the AKS Istio addon marks a significant advancement in simplifying and enhancing service mesh deployment and management. This feature provides substantial benefits, including improved performance, simplified configuration, and operational efficiency. We are committed to providing state-of-the-art features that empower our users to build resilient and scalable applications on AKS.

## Credits

This feature was made possible through the collaborative efforts of the AKS traffic team, the AKS Add-on team, and the Istio community. Special thanks to Niranjan Shankar, Shashank Barsin and all other AKS traffic team members; Robbie Zhang from the AKS CCP and Add-on team; and Brian Redmond for his guidance on this blog.

> For more detailed information, you can refer to the [Istio-based service mesh add-on for Azure Kubernetes Service](https://learn.microsoft.com/en-us/azure/aks/istio-native-sidecar). If you have any question with regard to this preview feature, please leave us an issue here: https://github.com/Azure/AKS/issues.

Reviewer 2: Removing locale from MS Learn URL

```suggestion
For more detailed information, you can refer to the [Istio-based service mesh add-on for Azure Kubernetes Service](https://learn.microsoft.com/azure/aks/istio-native-sidecar). If you have any question with regard to this preview feature, please leave us an issue here: https://github.com/Azure/AKS/issues.
```
